test/test_collector.py

```diff
@@ -66,6 +85,101 @@ def make_policy(env):
         raise NotImplementedError
 
 
+def _is_consistent_device_type(
+    device_type, policy_device_type, passing_device_type, tensordict_device_type
+):
+    if passing_device_type is None:
+        if device_type is None:
+            if policy_device_type is None:
+                return tensordict_device_type == "cpu"
+
+            return tensordict_device_type == policy_device_type
+
+        return tensordict_device_type == device_type
+
+    return tensordict_device_type == passing_device_type
+
+
+@pytest.mark.parametrize("num_env", [1, 3])
+@pytest.mark.parametrize("device", ["cuda", "cpu", None])
+@pytest.mark.parametrize("policy_device", ["cuda", "cpu", None])
+@pytest.mark.parametrize("passing_device", ["cuda", "cpu", None])
+@pytest.mark.skipif(torch.cuda.device_count() <= 1, reason="no cuda device found")
+def test_output_device_consistency(
+    num_env, device, policy_device, passing_device, seed=40
+):
+    if num_env == 1:
+
+        def env_fn(seed):
+            env = make_make_env("vec")()
+            env.set_seed(seed)
+            return env
+
+    else:
+
+        def env_fn(seed):
+            env = ParallelEnv(
+                num_workers=num_env,
+                create_env_fn=make_make_env("vec"),
+                create_env_kwargs=[{"seed": i} for i in range(seed, seed + num_env)],
+            )
+            return env
+
+    if policy_device is None:
+        policy = make_policy("vec")
+    else:
+        policy = ParametricPolicy().to(policy_device)
+
+    collector = SyncDataCollector(
+        create_env_fn=env_fn,
+        create_env_kwargs={"seed": seed},
+        policy=policy,
+        frames_per_batch=20,
+        max_frames_per_traj=2000,
+        total_frames=20000,
+        device=device,
+        passing_device=passing_device,
+        pin_memory=False,
+    )
+    for i, d in enumerate(collector):
+        assert (
+            True
+            if _is_consistent_device_type(
+                device, policy_device, passing_device, d.device.type
+            )
+            else False
+        )
+        if i == 5:
```

Comment:
I think we can stop earlier than that, one iteration is sufficient (i.e. you can break immediately)